classes/cloner.php

```diff
@@ -0,0 +1,159 @@
+     * @param object $data form data.
+     * @return array array containg the source and destination course categories.
+     */
+    public static function prepare($data) {
+        global $DB;
+
+        // Clone the category.
+
+        // Category selection.
+        $src = core_course_category::get($data->source);
+        $dest = core_course_category::get($data->destination);
+
+        // If a destination category name was supplied, create it and update the $dest object.
+        if (!empty($data->destcategoryname) && !empty($data->destcategoryidnumber)) {
+            if ($rec = $DB->get_record('course_categories', ['name' => trim($data->destcategoryname),
+                "idnumber" => trim($data->destcategoryidnumber), "parent" => $dest->id])) {
+                // We have an existing destination with all these details, use that one.
+                $dest = core_course_category::get($rec->id);
+            } else {
+                $dest = core_course_category::create([
+                    "name" => trim($data->destcategoryname),
+                    "idnumber" => trim($data->destcategoryidnumber),
+                    "parent" => $dest->id
+                ]);
+            }
+        }
+
+        return [$src, $dest];
+    }
+
+    /**
+     * Queues the cloning of the courses in the source category.
+     * @param object $data form data
+     * @param core_course_category $src source category
+     * @param core_course_category $dest destination category
+     */
+    public static function queue($data, core_course_category $src, core_course_category $dest) {
+        $courseids = array_keys($src->get_courses(['recursive' => false]));
+
+        foreach ($courseids as $courseid) {
+            $task = new clone_course_task();
+            $task->set_custom_data((object) [
+                'courseid' => $courseid,
+                'destid' => $dest->id,
+                'srcid' => $src->id,
+                'data' => $data
+            ]);
+            manager::queue_adhoc_task($task, true);
+        }
+    }
+
+    /**
+     * Performs the cloning of the given course to the destination category.
+     * This is intended to be called from within an adhoc task.
+     * @param int $courseid course to clone (in $src category)
+     * @param core_course_category $src source category
+     * @param core_course_category $dest destination category
+     * @param object $data form data
+     */
+    public static function clone_course(int $courseid, core_course_category $src, core_course_category $dest, object $data) {
+        global $DB;
+
+        core_php_time_limit::raise(600);
+
+        $course = get_course($courseid);
+        $idn = explode('_', $course->shortname);
+        $shortname = reset($idn) . '_' . $dest->idnumber;
+
+        // If a course matching the shortname and destination category already exists, skip it.
+        if ($DB->record_exists("course", ["shortname" => $shortname, "category" => $dest->id])) {
```

Comment:
this might be existing logic, but wouldn't we be better to match on the idnumber existing which is a proper unique id rather than its shortname which is a human readable thing and could be a bit floppy?

Reply:
it probably should just match the idnumber, and not even care about the category